credentials/credentials.go

```diff
@@ -158,11 +158,20 @@ func ParseSingleCredentials(credentialsMap map[string]interface{}) (Credentials,
 	default:
 		return nil, fmt.Errorf("entry %s: unknown credentials type: %s", id, credentialsType)
 	}
-	err := mapstructure.Decode(credentialsMap, credentials)
+	var validationErrors error
+	delete(credentialsMap, "type")
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		ErrorUnused: true,
+		Metadata:    nil,
+		Result:      credentials})
 	if err != nil {
+		validationErrors = multierror.Append(validationErrors, fmt.Errorf("entry %s: unable to create a decoder: %v", id, err))
+	}
+	if err := decoder.Decode(credentialsMap); err != nil {
 		return nil, fmt.Errorf("entry %s: invalid credentials data: %v", id, err)
```

Comment:
No part of the code you changed, but the error created here should be rather:

`fmt.Errorf("entry %s: invalid credentials data: %w", id, err)`

`%w` is used for wrapping an error and should be used instead of `%v` when the value is an error.